sycl/source/detail/queue_impl.hpp

```diff
@@ -419,6 +419,23 @@ class queue_impl {
   }
 
 private:
+  void finalizeHandler(handler &Handler, bool NeedSeparateDependencyMgmt,
+                       event &EventRet) {
+    if (MIsInorder) {
+      // Accessing and changing of an event isn't atomic operation.
```

Comment:
What does spec say about thread-safety of the in-order queue? If multiple threads concurrently submit commands to the same in-order queue then no order is guaranteed, right?

Reply:
The spec says:
> The in_order property adds the requirement that the SYCL queue provides in-order semantics where tasks are executed in the order in which they are submitted. Tasks submitted in this fashion can be viewed as having an implicit dependence on the previously submitted operation.

I believe, the following can be considered as correct.
When several threads attempt to submit a command simultaneously, the implementation should provide ordering between them.

Reply:
> the implementation should provide ordering between them.

But this would be implementation-defined ordering, right? So no clear in-order semantics.
Is this change fixing exactly the case where multiple threads submit simultaneously?
Was it just abnormally failing without this change?

Reply:
> But this would be implementation-defined ordering, right? So no clear in-order semantics.

Correct. Although, I can't see other ordering than impl-defined for the case of simultaneous calls.

> Is this change fixing exactly the case where multiple threads submit simultaneously?

Absolutely. Probably, there should be spin-lock instead of mutex in some cases.

> Was it just abnormally failing without this change?

I didn't observe such failures yet. This issue was found by eyeballing. I expect the issue is hard to reproduce due small size of non-atomic change.

Reply:
NIT: I suggest adding a comment explaining that this lock doesn't provide any specific order for the in-order queue, but just to served competing requests safely (in arbitrary order).